drivers/i2s/i2s_ll_stm32.c

```diff
@@ -908,6 +1047,8 @@ static const struct i2s_stm32_cfg i2s_stm32_config_##index = {		\
 	.i2s_clk_sel = CLK_SEL_##clk_sel,				\
 	.pcfg = PINCTRL_DT_DEV_CONFIG_GET(DT_NODELABEL(i2s##index)),	\
 	.irq_config = i2s_stm32_irq_config_func_##index,		\
+	.master_clk_sel =                                               \
+	 FOREACH_PORT(i2s##index##_mck, PORT_FUNC, PORT_RANGE)          \
```

Comment:
> A question for the sake of clarity, assuming `FOREACH_PORT` would be taking an instance number instead, would the internal usage of `DT_NODELABEL` within the subsequent macro calls be acceptable?

As long as the macro takes instances numbers (0, 1, ...) as input, that is fine.

Reply:
There are two ways I could think of to crack this nut without relying on additional non-mandatory entries in DT (e.g. label), note that the snippets below are just a prototype:

1. Create a shim to convert from instance numbers to node labels so that the existing recursive macros will remain as they are, the conversion would take form in a series of macros to cover all of the possible peripheral instances. E.g.

```
#if DT_NODE_HAS_STATUS(DT_NODELABEL(i2s1), okay)
#if DT_SAME_NODE(DT_NODELABEL(i2s1), DT_DRV_INST(0))
#define DT_NODE_INST_0 i2s1 
#elif DT_SAME_NODE(DT_NODELABEL(i2s1), DT_DRV_INST(1)) 
#define DT_NODE_INST_1 i2s1 
#elif DT_SAME_NODE(DT_NODELABEL(i2s1), DT_DRV_INST(2)) 
#define DT_NODE_INST_2 i2s1 
#elif DT_SAME_NODE(DT_NODELABEL(i2s1), DT_DRV_INST(3)) 
#define DT_NODE_INST_3 i2s1 
#elif DT_SAME_NODE(DT_NODELABEL(i2s1), DT_DRV_INST(4)) 
#define DT_NODE_INST_4 i2s1 
#else
#error "Unsupported instance number index when evaluating DT_DRV_INST to " \
           "DT_NODELABEL conversion for label i2s1"
#endif

/* Repeat above for all the supported peripherals, currently 5 */

#define INST_TO_NODE(id) DT_NODE_INST_ ## id

#define I2S_INIT(id)                    \
....                                    \
.master_clk_sel =                       \
         FOREACH_PORT(INST_TO_NODE(id)##_mck, PORT_FUNC, PORT_RANGE)  \

...

DT_INST_FOREACH_STATUS_OKAY(I2S_INIT)
```
2. Iterate over each of the phandles contained by `pinctrl-0` for each device instance and compare each with any of the possible node labels associated with the pins of interest (exhaustive list of available pin functions).

```
/ *
 * Add another level of iteration to cover each possible peripheral index to perform an exhaustive search.
 * This macro function will call FOREACH_PORT in a second pass as a recursive call similar to how it is now but
 * with extra arguments.
 * The function to evaluate would change to, where pin would be of the form i2sX_mck_pYZ.
 * X = [1, 5], Y = PORT_RANGE, Z = PIN_RANGE.
 * NODE_CHECK(node_id, pin) DT_SAME_NODE(node_id, DT_NODE_LABEL(pin))
 */ 
#define FOREACH_PERIPHERAL(node_id) ....

/* Iterate over each phandle 
#define FOREACH_FUN(id, prop, idx) \             
            FOREACH_PERIPHERAL(DT_INST_PROP_BY_IDX(id, prop, idx))

#define I2S_INIT(id)                    \
....                                    \
.master_clk_sel =                       \
         DT_INST_FOREACH_PROP_ELEM_SEP(id,   pinctrl_0, FOREACH_FUN, SEP); \

...

DT_INST_FOREACH_STATUS_OKAY(I2S_INIT)
```
The second option would be more expensive, as many times as available i2s peripherals (5), for the preprocessor. And in my opinion, not as readable as the first option; the iteration macros will get a bit more complicated. On the other hand, the conversion macros needed by option 1 may come across as a bit verbose with all the repetition added by each peripheral which unfortunately I don't think we could get away without.

I hope these options throw some light. Feel free to get in touch when you get to undertake the rework if that helped with the specifics.

Reply:
Regardless of performances, I have a strong preference to the second one. I'd hate having 100th of lines in drivers (not only I2S. As we said, this could be useful in other drivers and some have large number of instances: UART >  10 instances) just for the sake of being able to identify a pin. This kinda defeats the interest of instantiating the init macros.
Note that a simple alternative is to add a property to identify the pin presence (see https://github.com/zephyrproject-rtos/zephyr/pull/54946). Less exciting, but efficient (as long as user takes care of it) and readable.

Reply:
Point taken on mandraulic repetition, it would make sense to let the preprocessor do as much of the heavy lifting as possible.

I reckon I should incorporate an additional level to iterate through all the peripheral instances, that would integrate well with the existing incarnation of the driver and should become a bit of a stepping stone for the instance based rework in the near future.

Reply:
Reworked to generalise usage and make it more aligned with future plans aforementioned.

Building on [669f002](https://github.com/zephyrproject-rtos/zephyr/commit/669f00278037f3b7ed385e73777905bbf6947582), a DT_INSTANCE based usage would require:

```
/* Pin control function to check if the alternate function is enabled in DT */
#define NODE_CHECK(id, pin) COND_CODE_1(DT_NODE_EXISTS(DT_NODELABEL(pin)),     \
                                        DT_SAME_NODE(id, DT_NODELABEL(pin)),   \
                                        (false))

/* Range of peripheral devices */
#define DEV_RANGE 1, 2, 3, 4, 5

#define FUNC(inst, prop, idx) \
  FOREACH_DEV(DT_INST_PROP_BY_IDX(inst, prop, idx), i2s, mck, DEV_FUNC, DEV_RANGE)

#define I2S_INIT(inst)                                        \
...
 .master_clk_sel =                                            \
  DT_INST_FOREACH_PROP_ELEM_SEP(inst, pinctrl_0, FUNC, (SEP)) \
...

DT_FOREACH_STATUS_OKAY(I2S_INIT)

```